### ISSUE TITLE
Reflection/Introspection

### ISSUE THREAD

**Issue by dtornow** [servicemeshinterface/smi-spec#170]
Hello,

Are there any discussions/plans to add an introspection/reflection API to SMI? Currently, Service Mesh Resources exist in an "implicit context" of a Service Mesh. To identify/pinpoint a Service Mesh Instance one has to "hunt for clues".

For example, [Service Mesh Hub](https://github.com/solo-io/service-mesh-hub) scans for K8s Resources:

- [To detect an Istio instance](https://github.com/solo-io/service-mesh-hub/blob/1c196759e64fe418aacfd3858dbcbbe621f2e1a1/services/mesh-discovery/pkg/discovery/mesh/k8s/istio/istio_mesh_scanner.go#L62) 
(Specifically scanning for a Deployment with a Container with a name that includes "istiod")
- [To detect a Linkerd instance ](https://github.com/solo-io/service-mesh-hub/blob/1c196759e64fe418aacfd3858dbcbbe621f2e1a1/services/mesh-discovery/pkg/discovery/mesh/k8s/linkerd/linkerd_mesh_scanner.go#L64) 
(Specifically scanning for a Deployment with a Container with a name that includes "linkerd-io/controller")

Any thoughts on adding a Service Mesh Resource that should be created by the respective (SMI-compatible) Service Mesh Instance to make its presence discoverable?

This concept could further be extended to Services and Workloads

**Comment by grampelberg** [comment]
Hmm, what problem are you trying to solve?

**Comment by dtornow** [comment]
I would like to be able to discover Service Mesh Instances as well as the Services and Workloads the Service Mesh Instances know about on an SMI-level. 

Service Mesh Hub implements Mesh, Service, and Workload discovery but it hunts for Service Mesh Implementation specific clues.

**tl;dr**

Problem: Discover Service Mesh Instances, Services, and Workloads via SMI

**Comment by grampelberg** [comment]
So, services and workloads are discoverable via. Kubernetes. Why would we need to see those through SMI?

I assume by "service mesh instances" you're interested in knowing what's been installed on the cluster?

**Comment by dtornow** [comment]
> I assume by "service mesh instances" you're interested in knowing what's been installed on the cluster?

Yes. For example, I can install [two Istio instances on the same cluster](https://istio.io/blog/2018/soft-multitenancy/).

> So, services and workloads are discoverable via. Kubernetes. Why would we need to see those through SMI?

Individual Service Mesh implementations have individual rules which K8s Services and K8s Workloads they consider.

For example, Istio may restrict the Namespace to discover both Services and Workloads.

For another example, Linkerd [skips](https://github.com/linkerd/linkerd2/blob/10ecd8889e58f3ff9faee75479113c303fc7b5fa/controller/api/destination/endpoint_translator.go#L86) Workloads if unexpected errors occur.

**tl;dr**

A Service Mesh Implementations/Instances decides which Services and Workloads to include. Could we extend SMI to discover what the Service Mesh discovered?

**Comment by dtornow** [comment]
> Hmm, services are going to be extremely hard as a single service can have multiple meshes behind it.

At first glance, I don't expect much trouble here: I two Service Mesh Instances consider the same K8s Service, it'll still show up as two distinct Mesh Services belonging to two distinct Service Mesh Instances

We could take a page from Service Mesh Hub's playbook, SMH defines a Cluster, Mesh, MeshService and MeshWorkload CRD. Seems like a good starting point.

I see value add in both single as well as multi cluster scenarios, but yes, we do need to list the benefits. 

> Either way, I think we should flesh out a common status for the resources.

Didn't totally get that
